File: 322-coin-change/322-coin-change.ts
// https://www.youtube.com/watch?v=H9bfqozjoqs

function coinChange(coins: number[], amount: number): number {
  
  const N = coins.length,
        map = new Array(amount + 1).fill(amount + 1);
  
  map[0] = 0;
  
  for(let i = 1; i <= amount; i++)
    for(let coin of coins)
      if(coin <= i){
        // console.log({i, coin, m_i: map[i], m_i_c: map[i-coin]})
        map[i] = Math.min(map[i], 1 + map[i - coin]);        
      }

  // console.log(map);
  return map[amount] > amount ? -1 : map[amount]; 
}

File: 322-coin-change/322-coin-change.java
class Solution {
	public int coinChange(int[] coins, int amount) {
		int N = coins.length;
		int[] dp = new int[1 + amount];

		Arrays.sort(coins);
		Arrays.fill(dp, 1 + amount);

		dp[0] = 0;

		for (int i = 1; i <= amount; i++) {

			for (int j = 0; j < N && coins[j] <= i; j++) {
				dp[i] = Math.min(1 + dp[i - coins[j]], dp[i]);
			}
		}
    
		return dp[amount] > amount ? -1 : dp[amount];
	}
}